lib/reservation_manager.rb

```diff
@@ -0,0 +1,117 @@
+require_relative "room"
+require_relative "reservation"
+require "pry"
+
+module Hotel
+  class ReservationManager
+    attr_reader :rooms, :reservations
+
+    def initialize
+      @rooms = (1..20).map { |num| Hotel::Room.new(number: num) }
+      @reservations = []
+    end
+
+    # creates a reservation for first available room if there are any available
+    def request_reservation(check_in, check_out, block_name: nil, booking_name:, discount: nil)
+      # checks for available rooms
+      rooms = available_rooms(check_in, check_out)
+
+      # raises an ArgumentError if no rooms are available
+      raise ArgumentError, "No available rooms" if rooms.length == 0
+
+      # creates a reservation
+      reservation = Hotel::Reservation.new(check_in: check_in,
+                                           check_out: check_out,
+                                           room: rooms.first,
+                                           block_name: block_name,
+                                           booking_name: booking_name,
+                                           discount: discount)
+
+      # adds reservation to list of all reservations
+      reservations << reservation
+
+      # add reservation to corresponding room
+      rooms.first.add_reservation(reservation)
+      return reservation
+    end
+
+    # returns a list of all reservations for a given date
+    def reservations_by_date(date)
+      reservations.find_all { |reservation| reservation.all_dates.include?(Date.parse(date)) }
+    end
+
+    # returns a list of all available rooms for a given date range
+    def available_rooms(check_in, check_out)
+      check_in = Date.parse(check_in)
+      check_out = Date.parse(check_out)
+
+      if check_in == check_out
+        booking_dates = [check_in]
+      else
+        booking_dates = (check_in..check_out).to_a
+      end
+
+      # returns a list of all rooms, if no reservations have been made
+      return rooms if reservations.empty?
```

Comment:
This is probably unnecessary since if `reservations` is empty it will still return all the rooms.